src/fitting.jl

```diff
@@ -0,0 +1,139 @@
+import ITensorNetworks as itn
+import ITensors as it
+import NamedGraphs.PartitionedGraphs as npg
+using NamedGraphs: NamedEdge
+using Printf
+
+using ITensorNetworks: ITensorNetwork, induced_subgraph, ket_vertices, rename_vertices, inv_vertex_map
+
+
+@kwdef mutable struct FittingProblem{State,OverlapNetwork}
+  state::State
+  overlapnetwork::OverlapNetwork
+  overlap::Number = 0
+end
+
+overlap(F::FittingProblem) = F.overlap
+overlapnetwork(F::FittingProblem) = F.overlapnetwork
+state(F::FittingProblem) = F.state
+
+function set(
+  F::FittingProblem;
+  state=state(F),
+  overlapnetwork=overlapnetwork(F),
+  overlap=overlap(F),
+)
+  return FittingProblem(; state, overlapnetwork, overlap)
+end
+
+function extracter!(problem::FittingProblem, prev_region, region; kws...)
+  tn = state(problem)
+  path = itn.edge_sequence_between_regions(tn, prev_region, region)
+  tn = itn.gauge_walk(itn.Algorithm("orthogonalize"), tn, path)
+  verts = unique(vcat(src.(path), dst.(path)))
+  factors = [tn[v] for v in verts]
+  o_tn = overlapnetwork(problem)
+  o_tn = itn.update_factors(o_tn, Dict(zip(verts, factors)))
+  pe_path = npg.partitionedges(itn.partitioned_tensornetwork(o_tn), path)
+  o_tn = itn.update(
+    itn.Algorithm("bp"), o_tn, pe_path; message_update_function_kwargs=(; normalize=false)
+  )
+
+  problem.state = tn
+  problem.overlapnetwork = o_tn
+
+  local_tensor = itn.environment(o_tn, region)
+  sequence = itn.contraction_sequence(local_tensor; alg="optimal")
+  local_tensor = dag(it.contract(local_tensor; sequence))
+
+  return local_tensor
+end
+
+function fitting_inserter!(problem::FittingProblem, local_tensor, region; kws...)
+  inserter!(problem, local_tensor, region; kws...)
+  on = overlapnetwork(problem)
+  tn = state(problem)
+  on = itn.update_factors(
+    on, Dict(zip(region, [dag(tn[v]) for v in region]))
+  )
+  problem.overlapnetwork = on
+  return nothing
+end
+
+function region_iterator_action!(
+  problem::FittingProblem;
+  region,
+  prev_region=nothing,
+  extracter_kwargs=(;),
+  updater_kwargs=(;),
+  inserter_kwargs=(;),
+  sweep,
+  kwargs...,
+)
+  prev_region = isnothing(prev_region) ? collect(vertices(state(problem))) : prev_region
```

Comment:
Interesting that you needed `prev_region` too. I was finding I needed it for the DMRG subspace expansion ("post" subspace expansion scheme) so it's made me want to think about if there is cleaner and more 'universal' way to make it available to different problem implementations.

Reply:
Related to that, I found I was able to obtain the previous region by inspecting the cache (in my case operator), like this:
```
prev_region = setdiff(itn.pos(operator(problem)), region)
```
Could a pattern like that (moved inside one of the extracter, updater, inserter functions) work for your cases?

Reply:
One other option is I keep it inside the `FittingProblem` type itself

Reply:
I like that idea of storing it in the problem for now. It seems like something that in the future may call for using a more advanced cache or state type that does store that info.

Reply:
Great, I have stored it inside the problem itself and that seems to work nicely.